### PR TITLE
Rpc wrapper autogeneration

### PR DESCRIPTION
**Pull request by roman-khimov**
### Problem

#2705.

### Solution

Reuse existing infrastructure as much as possible and create something emitting code that can be compiled. To be extended.

### CODE REVIEW

File: pkg/smartcontract/rpcbinding/binding.go [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+		meths = dropManifestMethods(meths, std.Optional)
+	}
+	if std.Base != nil {
+		return dropStdMethods(meths, std.Base)
+	}
+	return meths
+}
+
+func scTypeToGo(name string, typ smartcontract.ParamType, overrides map[string]binding.Override) (string, string) {
+	// TODO: remap structures
```
Comment (AnnaShaleva) [inline pkg/smartcontract/rpcbinding/binding.go:121]: This one can be removed?

**Comment by AnnaShaleva** [comment]
And we need to adapt tests for Windows and its line endings (`\r\n\t` vs `\n\t`).

File: pkg/smartcontract/rpcbinding/binding.go [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+import (
+{{range $m := .Imports}}	"{{ $m }}"
+{{end}})
+
+// Hash contains contract hash.
+var Hash = {{ .Hash }}
+
+// Invoker is used by ContractReader to call various safe methods.
+type Invoker interface {
+	{{if or .IsNep11D .IsNep11ND}}nep11.Invoker{{end}}
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/smartcontract/rpcbinding/binding.go:41]: I don't like the way it looks in the template, but it makes the output look better, so let's try it this way.

File: cli/smartcontract/generate.go [hunk 3]
```diff
@@ -7,39 +7,59 @@ import (
...
-		},
-	},
+	Flags:       generatorFlags,
+}
+
+var generateRPCWrapperCmd = cli.Command{
+	Name:        "generate-rpcwrapper",
+	Usage:       "generate RPC wrapper to use for data reads",
+	UsageText:   "neo-go contract generate-rpcwrapper --manifest <file.json> --out <file.go> --hash <hash>",
+	Description: ``,
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply cli/smartcontract/generate.go]: Dropped.

File: pkg/smartcontract/rpcbinding/binding.go [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+)
+
+const srcTmpl = `
+{{- define "METHOD" -}}
+// {{.Name}} {{.Comment}}
+func (c *ContractReader) {{.Name}}({{range $index, $arg := .Arguments -}}
+	{{- if ne $index 0}}, {{end}}
+		{{- .Name}} {{.Type}}
+	{{- end}}) {{if .ReturnType }}({{ .ReturnType }}, error) {
+	return unwrap.{{.CallFlag}}(c.invoker.Call(Hash, "{{ .NameABI }}"
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/smartcontract/rpcbinding/binding.go]: Fixed.

File: pkg/smartcontract/rpcbinding/binding.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/smartcontract/rpcbinding/binding.go]: Sure.

File: docs/compiler.md [hunk 5]
```diff
@@ -410,6 +410,30 @@ given RPC server and wallet and paying 0.00001 extra GAS for this transaction):
...
+```
+
+Notice that some structured types can be omitted this way (when a function
+returns some structure it's just an "Array" type in the manifest with no
+internal details), but if the contract you're using is written in Go
+originally you can create a specific configuration file during compilation
+that will add this data for wrapper generator to use:
+
+```
+$ ./bin/neo-go contract compile -i contract.go --config contract.yml -o contract.nef --manifest manifest.json --bindings contract.bindings.yml
```
Comment (AnnaShaleva) [inline docs/compiler.md:433]: We probably need to add an example of configuration file.

File: docs/compiler.md [hunk 6]
```diff
@@ -434,6 +434,19 @@ $ ./bin/neo-go contract compile -i contract.go --config contract.yml -o contract
...
+To simplify interacting with the contract via RPC you can generate
+contract-specific RPC bindings with the "generate-rpcwrapper" command. At the
+moment it only works for safe (read-only) methods. If your contract is NEP-11
+or NEP-17 that's autodetected and an appropriate package is included as
+well. Notice that the type data available in the manifest is limited, so in
+some cases the interface generated may use generic stackitem types. Iterators
+are not supported yet.
+
+```
+$ ./bin/neo-go contract generate-rpcwrapper --manifest manifest.json --out rpcwrapper.go --hash 0x1b4357bff5a01bdf2a6581247cf9ed1e24629176
```
Comment (AnnaShaleva) [inline docs/compiler.md:447]: Add a couple of words about bindings config as far?

File: docs/compiler.md (hunk 5 above)
Comment (roman-khimov) [reply docs/compiler.md:433]: It's supposed to be generated.

File: docs/compiler.md (hunk 6 above)
Comment (roman-khimov) [reply docs/compiler.md:447]: It's not used at the moment, so it's intentionally omitted.

File: docs/compiler.md (hunk 5 above)
Comment (AnnaShaleva) [reply docs/compiler.md:433]: I've created it manually for the contract that wasn't compiled by me :D

File: docs/compiler.md (hunk 6 above)
Comment (AnnaShaleva) [reply docs/compiler.md:447]: At least `Overrides` are taken from it by `TemplateFromManifest`.

Comment (roman-khimov) [reply docs/compiler.md:447]: Yeah, but it's an internal thing.

**Review by AnnaShaleva — APPROVED** [review]

**Pull request merged**